management/src/main/scala/akka/management/AkkaManagement.scala

```diff
@@ -0,0 +1,186 @@
+      val protocol = if (_connectionContext.isSecure) "https" else "http"
+      val selfBaseUri = Uri(s"$protocol://$hostname:$port/${settings.Http.BasePath}")
+      val providerSettings = ManagementRouteProviderSettingsImpl(selfBaseUri)
+
+      val combinedRoutes: Try[Route] = prepareCombinedRoutes(settings, providerSettings)
+
+      combinedRoutes match {
+        case Success(routes) ⇒
+          // TODO instead of binding to hardcoded things here, discovery could also be used for this binding!
+          // Basically: "give me the SRV host/port for the port called `akka-bootstrap`"
+          // discovery.lookup("_akka-bootstrap" + ".effective-name.default").find(myaddress)
+          // ----
+          // FIXME -- think about the style of how we want to make these available
+
+          val serverFutureBinding =
+            Http().bindAndHandle(
+              RouteResult.route2HandlerFlow(routes),
+              hostname,
+              port,
+              connectionContext = this._connectionContext
+            )
+
+          serverBindingPromise.completeWith(serverFutureBinding).future.flatMap { _ =>
+            log.info("Bound Akka Management (HTTP) endpoint to: {}:{}", hostname, port)
+            selfUriPromise.success(selfBaseUri).future
+          }
+
+        case Failure(ex) ⇒
+          log.warning(ex.getMessage)
+          Future.failed(new IllegalArgumentException("Failed to start Akka Management HTTP endpoint.", ex))
+      }
+    } else selfUriPromise.future
+  }
+
+  private def prepareCombinedRoutes(managementSettings: AkkaManagementSettings,
+                                    providerSettings: ManagementRouteProviderSettings): Try[Route] = {
+    val basePath: Directive[Unit] = {
+      val pathPrefixName = settings.Http.BasePath.getOrElse("")
+      if (pathPrefixName.isEmpty) rawPathPrefix(pathPrefixName) else pathPrefix(pathPrefixName)
+    }
+
+    def wrapWithAuthenicatorIfPresent(inner: Route): Route =
+      _asyncAuthenticator match {
+        case Some(asyncAuthenticator) ⇒
+          authenticateBasicAsync[String](realm = "secured", asyncAuthenticator)(_ ⇒ inner)
+        case _ ⇒ inner
+      }
+
+    val combinedRoutes = routeProviders.map { provider =>
+      log.info("Including HTTP management routes for {}", Logging.simpleName(provider))
+      provider.routes(providerSettings)
+    }
+
+    Try {
+      if (combinedRoutes.nonEmpty) {
+        basePath {
+          wrapWithAuthenicatorIfPresent(Directives.concat(combinedRoutes: _*))
+        }
+      } else
+        throw new IllegalArgumentException(
+            "No routes configured for akka management! " +
+            "Double check your `akka.management.http.route-providers` config.")
+    }
+  }
+
+  def stop(): Future[Done] =
+    if (bindingFuture.get() == null) {
+      Future.successful(Done)
+    } else {
+      val stopFuture = bindingFuture.get().flatMap(_.unbind()).map(_ => Done)
```

Comment:
Shouldn't this be a cas as well - extra weird that start() can be called multiple times, but stop() could throw an NPE here if called multiple times

Reply:
good call